### DISCUSSION TITLE
Unexpected NotNullViolation with Sqlite and custom ToSql implementation

### DISCUSSION THREAD

**Discussion by PhotonQuantum** [diesel-rs/diesel#3421] [Q&A]
I have a custom rust type like this
```rust
#[derive(Debug, FromSqlRow, AsExpression)]
#[diesel(sql_type = Text)]
struct CustomWrapper(String);
```

... and its `ToSql` implementation

```rust
impl ToSql<Text, Sqlite> for CustomWrapper {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
        // Change this to `out.set_value(&self.0);` works.
        out.set_value(self.0.to_string());
        Ok(serialize::IsNull::No)
    }
}
```

I expect it to work on a `TEXT NOT NULL` field. However, when inserting it into the table, I got `DatabaseError(NotNullViolation, "NOT NULL constraint failed: kvs.v")`. And if I feed set_value a borrowed str, it works as intended.

I read the doc and the implementation of [`SqliteBindValue`](http://docs.diesel.rs/2.0.x/diesel/sqlite/struct.SqliteBindValue.html) but still don't know what went wrong. Could you please help me with this?

The minimal reproduce repo is at https://github.com/PhotonQuantum/diesel-not-null-test.

**Comment by weiznich** [comment]
I can reproduce this problem. This seems to be a bug in diesel. I've opened https://github.com/diesel-rs/diesel/issues/3425 based on your report to track this issue. I will add additional details there.
